### ISSUE TITLE
Opening LevelDB with corrupted MANIFEST-0* file

### ISSUE THREAD

**Issue by dasasd122311** [facebook/rocksdb#12331]
I am new to using RocksJava. I am trying to open LevelDB database through RocksJava. When I open it, I get an error - MANIFEST file is corrupted. Is there any way I can restore MANIFEST files in all databases or at least open the database without using the Manifest file to display all keys and values?

```
ERROR: bad block contents  The file C:\Users\db_path/MANIFEST-004651 may be corrupted.
ERROR: bad block contents  The file C:\Users\db_path1/MANIFEST-000075 may be corrupted
```

I would like to open LevelDB databases using RocksDB.openReadOnly() to use the data without modifying the database files

**Comment by alanpaxton** [comment]
Is the manifest really corrupted, or are you just hitting a compatibility issue ? I am not sure how compatible RocksDB and LevelDB formats are; you could try opening using RocksJava with the oldest RocksDB format version
```
  public void openWithOldestFormatVersion() {
    final BlockBasedTableConfig blockBasedTableConfig =
        new BlockBasedTableConfig().setFormatVersion(0);

    try (final Options options = new Options().setTableFormatConfig(blockBasedTableConfig);
         final RocksDB ignored = RocksDB.openReadOnly(options, "<Your DB path here>")) {
    } catch (RocksDBException e) {
      // Access the DB
    }
  }
```
If you want to dig deeper see [Docs](https://github.com/facebook/rocksdb/wiki/Rocksdb-BlockBasedTable-Format)

If the above doesn't work, you may be able to migrate your DB by opening it using LevelDB, iterating the keys and values, and writing to a new RocksDB database.

**Comment by dasasd122311** [comment]
> Действительно ли манифест поврежден, или у вас просто возникла проблема совместимости? Я не уверен, насколько совместимы форматы RocksDB и LevelDB; вы можете попробовать открыть с помощью RocksJava самую старую версию формата RocksDB.
> 
> ```
>   public void openWithOldestFormatVersion() {
>     final BlockBasedTableConfig blockBasedTableConfig =
>         new BlockBasedTableConfig().setFormatVersion(0);
> 
>     try (final Options options = new Options().setTableFormatConfig(blockBasedTableConfig);
>          final RocksDB ignored = RocksDB.openReadOnly(options, "<Your DB path here>")) {
>     } catch (RocksDBException e) {
>       // Access the DB
>     }
>   }
> ```
> 
> Если вы хотите копнуть глубже, см. [Документы .](https://github.com/facebook/rocksdb/wiki/Rocksdb-BlockBasedTable-Format)
> 
> Если вышеперечисленное не сработает, вы можете перенести свою БД, открыв ее с помощью LevelDB, повторив ключи и значения и записав в новую базу данных RocksDB.

I think the problem is really due to LevelDB and RocksDB incompatibility, but setting it to the old database format didn't solve the problem. My LevelDB databases have been around since 2022

**Comment by alanpaxton** [comment]
It was worth a try. I guess you will need to write copy code in C++, as there isn't a Java API for LevelDB.

**Comment by dasasd122311** [comment]
Yeah, I'm closing the issue

**Comment by adamretter** [comment]
@dasasd122311 There was a Java port of LevelDB (I haven't checked if the storage format is the same) - might be helpful for you - https://github.com/dain/leveldb

**Comment by dasasd122311** [comment]
> @dasasd122311 There was a Java port of LevelDB (I haven't checked if the storage format is the same) - might be helpful for you - https://github.com/dain/leveldb

I use its fork - https://github.com/halibobor/leveldb-java. Because the fork has been updated and has no problems reading ldb files. But the code uses the library org.iq80.snappy. I got EXCEPTION_ACCESS_VIOLATION error. I read that the owner of the given wrote that they moved to the new aircompressor library, but I don't know how I can rewrite the LevelDB library code, because some methods are used, some are removed and stuff. I thought switching to RocksDB would solve my problem because it updates frequently, but RocksDB is incompatible with my database
